### ISSUE TITLE
Feature/3.0 request: Hide Adapter from implementers

### ISSUE THREAD

**Issue by NightlyNexus** [f2prateek/rx-preferences#67]
Users can't do anything special with the SharedPreferences or Editor except store strings.
Instead of an Adapter, how about making the Adapter internal and exposing a Converter:
```java
public interface Converter<T> {
  @NonNull T deserialize(@NonNull String serialized);
  @NonNull String serialize(@NonNull T value);
}
```
Internally, the adapter delegates to the supplied Converter.
```java
final class CustomTypeStringAdapter<T> implements RealPreference.Adapter<T> {
  private final Converter<T> converter;
  
  CustomTypeStringAdapter(Converter<T> converter) {
    this.converter = converter;
  }
  
  @Override public T get(@NonNull String key,
      @NonNull SharedPreferences preferences) {
    String string = preferences.getString(key, null);
    if (string == null) {
      throw new AssertionError(); // Not called unless key is present.
    }
    return converter.deserialize(string);
  }
  
  @Override public void set(@NonNull String key, @NonNull T value,
      @NonNull SharedPreferences.Editor editor) {
    String serialized = converter.serialize(value);
    checkNotNull(serialized,
        "Serialized string must not be null from value: " + value);
    editor.putString(key, serialized);
  }
}
```

**Comment by PaulWoitaschek** [comment]
I'm actually doing that now. I have a generic method that creates a preference using a moshi `JsonAdapter`: 

```java
  @NonNull private <T> Preference<T> adapterPreference(String key, @Nullable T defaultValue, JsonAdapter<T> adapter) {
    return prefs.getObject(key, defaultValue, new Preference.Adapter<T>() {
      @Override public T get(@NonNull String key, @NonNull SharedPreferences preferences) {
        String value = preferences.getString(key, null);
        try {
          assert value != null; // wont be called
          return adapter.fromJson(value);
        } catch (IOException e) {
          throw new RuntimeException(e);
        }
      }

      @Override public void set(@NonNull String key, @NonNull T value, @NonNull SharedPreferences.Editor editor) {
        String stringValue = adapter.toJson(value);
        editor.putString(key, stringValue);
      }
    });
  }
```

**Comment by NightlyNexus** [comment]
Yep. Even more generically, the `CustomTypeStringAdapter` could be user code.
The main advantage of pushing the Adapter as an internal implementation is hiding SharedPreferences from users (making it obvious not to call `.apply()` and just making the API slightly simpler).
Tracking here: https://github.com/f2prateek/rx-preferences/pull/75

**Comment by f2prateek** [comment]
Closed with #75
